### PR TITLE
[rush-sdk] supports to be called in script

### PR DESCRIPTION
**Pull request by chengcyber**
## Summary

Make `@rushstack/rush-sdk` can be called in script. which is a scenario has not been implemented.

## Details

1. find `rush.json` up
2. load `rush.json` to get `rushVersion`
3. get install-run folder by `common/temp/install-run/@microsoft+rush@${rushVersion}`
4. try resolve `rush-lib` the first time. If no errors, DONE
5. invoke `node common/scripts/install-run-rush.js --help`
6. try resolve `rush-lib` the second time

## How it was tested

Local

### CODE REVIEW

File: libraries/rush-sdk/src/index.ts
Comment (iclanton) [inline libraries/rush-sdk/src/index.ts:74]: This functionality already exists as `RushConfiguration.tryFindRushJsonLocation`. Can we just reuse that?

File: libraries/rush-sdk/src/index.ts
Comment (iclanton) [inline libraries/rush-sdk/src/index.ts:96]: This is repeated on line 104. Stick this in a helper function.

**Comment by chengcyber** [comment]
Hi @iclanton , changes applied, please take another look~

- remove `find-up`, use the functionality in `RushConfiguration`
- put `require` rush-lib in one place
- add a test case for scenario 3

File: libraries/rush-sdk/src/index.ts
Comment (chengcyber) [reply libraries/rush-sdk/src/index.ts]: Unfortunately no, since the task of `rush-sdk` is to load `rush-lib`, it can not reuse logic inside rush-lib. but i can remove the dependency of `find-up` and copy paste the logic to `rush-sdk`. hopes it makes thing better.

File: libraries/rush-sdk/src/index.ts
Comment (iclanton) [inline libraries/rush-sdk/src/index.ts:139]:
```suggestion
/**
 * @privateRemarks
 * Keep this in sync with `RushConfiguration.tryFindRushJsonLocation`.
 */
function findRushJsonLocation(startingFolder: string): string | undefined {
```

Can you add a similar note in the doc comment of `RushConfiguration.tryFindRushJsonLocation`?

File: libraries/rush-sdk/src/test/script.test.ts
Comment (iclanton) [inline libraries/rush-sdk/src/test/script.test.ts:15]: This test will get broken when the `rush-lib` top-level API changes.

Is this test running `install-run-rush.js`? Tests shouldn't be making network requests.

File: libraries/rush-sdk/src/index.ts
Comment (chengcyber) [reply libraries/rush-sdk/src/index.ts:139]: Sure

File: libraries/rush-sdk/src/test/script.test.ts
Comment (chengcyber) [reply libraries/rush-sdk/src/test/script.test.ts:15]: Got it, let me use a mock rush-lib exports instead.

**Comment by chengcyber** [comment]
Hi @iclanton , changes applied

- Add remarks on helper functions
- Add mock `rush-lib` when testing `rush-sdk`
  - no longer need network
  - no failing when `rush-lib` API changes

File: libraries/rush-sdk/src/test/script.test.ts
Comment (iclanton) [inline libraries/rush-sdk/src/test/script.test.ts:7]:
```suggestion
import * as path from 'path';
import { Executable } from '@rushstack/node-core-library';

const rushSdkPath: string = path.join(__dirname, '../../lib/index.js');
const sandboxRepoPath: string = `${__dirname}/sandbox`);
const mockPackageFolder: string = `${sandboxRepoPath}/mock-package`);
const mockRushLibPath: string = `${__dirname}/fixture/mock-rush-lib.js`);
```

File: libraries/rush-sdk/src/index.ts
Comment (iclanton) [inline libraries/rush-sdk/src/index.ts:148]:
```suggestion
function tryFindRushJsonLocation(startingFolder: string): string | undefined {
```

File: libraries/rush-sdk/src/index.ts
Comment (iclanton) [inline libraries/rush-sdk/src/index.ts:68]:
```suggestion
    const rushJsonPath: string | undefined = tryFindRushJsonLocation(process.cwd());
```

File: common/changes/@microsoft/rush/feat-rush-sdk-s3_2021-12-14-17-01.json
Comment (iclanton) [inline common/changes/@microsoft/rush/feat-rush-sdk-s3_2021-12-14-17-01.json:5]:
```suggestion
      "comment": "Add support for directly invoking a script that depends on `rush-sdk` from inside a Rush repo.",
```

Comment (iclanton) [reply common/changes/@microsoft/rush/feat-rush-sdk-s3_2021-12-14-17-01.json:5]: Rush change.

File: libraries/rush-sdk/src/index.ts
Comment (iclanton) [inline libraries/rush-sdk/src/index.ts:100]: There should probably be some `global.__rush__howRushLibModuleWasFound` property that gets set in each scenario.

**Review by iclanton — APPROVED** [review]

File: libraries/rush-sdk/src/index.ts
Comment (elliot-nelson) [inline libraries/rush-sdk/src/index.ts:56]:
```suggestion
```

**Review by elliot-nelson — APPROVED** [review]

**Comment by elliot-nelson** [comment]
I'm excited about this one, in our monorepo I've got my eye on about 15 places where I could rip out `rush-lib` and replace with `rush-sdk` instead.

File: libraries/rush-sdk/src/index.ts
Comment (octogonz) [inline libraries/rush-sdk/src/index.ts:152]: `10` is a weird limit. Apparently I wrote this code way back in 4fcb4698d8065df10a3061d0d8449ca7650736e6, but I have no idea why. 😁  If you search this code base for the regex `parent.*path.dirname` there are a number of other similar loops that don't have this limit.

(It's out of scope for this PR but I'm tempted to fix it anyway.)

Comment (octogonz) [reply libraries/rush-sdk/src/index.ts:152]: Heft also has a copy+paste of this algorithm in [RushUtilities._findRushJsonFolder()](https://github.com/microsoft/rushstack/blob/a13865bef9a20dab28c044be3504c7326bfe94b1/apps/heft/src/utilities/RushUtilities.ts#L12)

Comment (octogonz) [reply libraries/rush-sdk/src/index.ts:152]: Some of them are also doing recursion instead of loops which wastes stack.  

👉 After this PR is merged, I can make a separate PR to clean it up.

Comment (iclanton) [reply libraries/rush-sdk/src/index.ts:152]: Why bother having a limit at all?

```suggestion
  while (true) {
```

File: libraries/rush-sdk/src/index.ts
Comment (octogonz) [inline libraries/rush-sdk/src/index.ts:110]:
```ts
  // to the devDependencies for your project.
  throw new Error('The "@rushstack/rush-sdk" package context has not been initialized.');
```

This error message is only relevant to Scenario 1.  But Scenario 3 is now the most common/important error case.

Comment (octogonz) [reply libraries/rush-sdk/src/index.ts:110]: Suggested new text:

```ts
'The @rushstack/rush-sdk package was not able to load the Rush engine:'
```

And then we need to show different strings depending on what went wrong, something like:

- `'Unable to find rush.json in the current folder or its parent folders. This tool is meant to be invoked from a working directory inside a Rush repository.'`
- `'The @microsoft/rush-lib package failed to install'`
- `'The @microsoft/rush-lib package failed to load'`

File: libraries/rush-sdk/src/index.ts
Comment (octogonz) [inline libraries/rush-sdk/src/index.ts:89]: This output is important, however I agree that it may be rather noisy on the console.

Maybe what we could do is save it to a text string, and then print it to `console.error()` only if the operation fails.

File: libraries/rush-sdk/src/index.ts
Comment (octogonz) [inline libraries/rush-sdk/src/index.ts:87]: Installation can potentially take a long time, like 30s.  We should do something like:

```
console.log('The Rush engine has not been installed yet. Invoking install-run-rush.js...')
```

File: libraries/rush-sdk/src/index.ts
Comment (octogonz) [inline libraries/rush-sdk/src/index.ts:84]: Ideally we should have a switch (e.g. `RUSH_SDK_DEBUG=1` environment variable) that enables verbose logging of each step of this algorithm, for example if it is failing on a CI machine where we can't just attach a debugger.

However I think it's outside the scope of this PR. It should be a separate PR.

File: libraries/rush-sdk/src/index.ts
Comment (octogonz) [reply libraries/rush-sdk/src/index.ts:152]: I'm agreeing with you; just pointing out that the associated problems involve 5 or 6 files in this repo and should be a separate PR

File: libraries/rush-sdk/src/index.ts
Comment (octogonz) [inline libraries/rush-sdk/src/index.ts:126]: When the above code fails, it does `throw new Error()`.

I envisioned that the CLI tool might look like this:
```ts
import { RushConfiguration } from  '@rushstack/rush-sdk';

console.log('Doing something useful...');
```

In which case the `import` fails with an uncaught exception:
```ts
C:\MyRepo>helpful-tool
rush-sdk\lib\index.js:104
    throw new Error('The "@rushstack/rush-sdk" package context has not been initialized.');
    ^

Error: The "@rushstack/rush-sdk" package context has not been initialized.
    at Object.<anonymous> (C:\Git\rushstack\libraries\rush-sdk\lib\index.js:104:11)        
    at Module._compile (internal/modules/cjs/loader.js:1068:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1097:10)
```

This isn't great however.  There's a couple ways we could improve it:

1. `rush-sdk` does `console.error()` and then `process.exit(1)`.  Generally `process.exit()` should be avoided but it may be okay here.
2. The caller must call `try { require('rush-sdk') }` in the entry point for their tool.  They can also use regular `import { ... } from 'rush-sdk'` in other files, but they MUST ensure that those imports are evaluated after the entry point has loaded the SDK.

Which is better?

**Comment by octogonz** [comment]
@chengcyber thank you for implementing this!  It is an important feature.  

My feedback was mainly about the user experience for a person who is invoking a CLI tool that imports `rush-sdk`, in various error cases.  We don't need to make it perfect in this PR -- I'm fine with deferring some of the ideas to a later PR.

IMO this is the most important flow:
```
# Install a CLI tool that uses rush-sdk
$ npm install -g @my-company/helpful-tool

# I'm trying to invoke the tool outside of a Rush folder
$ helpful-tool

Error: The @rushstack/rush-sdk package was not able to load the Rush engine:
Unable to find rush.json in the current folder or its parent folders. 
This tool is meant to be invoked from a working directory inside a Rush repository.

$ cd git/my-repo

# It takes 20 seconds to install, give the user some idea what's up
$ helpful-tool

The Rush engine has not been installed yet. Invoking install-run-rush.js...

Welcome to Helpful Tool!
usage: helpful-tool  [-h] [-d] <command> ...
```

**Comment by elliot-nelson** [comment]
Something to keep in mind (not for _this_ PR, but maybe future ones) is that the whole stdout/stderr thing in Rush makes it hard to create machine-usable scripts with this new library.

For example, I create a script for my repo that is supposed to print JSON to stdout; as soon as I invoke this line:

```js
import { RushConfiguration } from '@rushstack/rush-sdk';
```

I have to assume my script is useless because it could be filled with 20 lines of rush-run-install spam.

(Not much you can do about this except document it, at this point... it would need a bigger revamp to ensure that all of that output went safely to stderr instead, which would then possibly trigger rush "failures" due to stderr usage, etc. etc.)

**Comment by chengcyber** [comment]
Hi @octogonz, @elliot-nelson, thanks to your valuable suggestions!

### Changes applied as follow:

1. a new `global.___rush___rushLibModuleFromInstallAndRunRush` to track rush-lib loaded from `install-run-rush.js`
2. logging improvement. Error message collected in Scenario 3 will be printed by `console.error` instead of `throw new Error` before `process.exit(1)`
3. `RUSH_SDK_DEBUG=1` turns on verbose logging
4. stderr of the `install-run-rush` spawned process is collected as well and is printed when error occurs.

```
~/workspace 6s                                                                                                                 12:12:55 AM
❯ node index.js
Unable to find rush.json in the current folder or its parent folders. This tool is meant to be invoked from a working directory inside a Rush repository.
```

```
~/workspace/rushstack2 test-rush-plugins*                                                                                      12:40:35 AM
❯ RUSH_SDK_DEBUG=1 my-cli


Rush Multi-Project Build Tool 5.57.1 - https://rushjs.io
Node.js version is 14.17.3 (LTS)


Trying to acquire lock for pnpm-6.7.1
Acuired lock for pnpm-6.7.1
Found pnpm version 6.7.1 in /home/cheng/.rush/node-v14.17.3/pnpm-6.7.1

Symlinking "/home/cheng/workspace/rushstack2/common/temp/pnpm-local"
  --> "/home/cheng/.rush/node-v14.17.3/pnpm-6.7.1"
Acquiring lock for "common/autoinstallers/command-plugins" folder...
Autoinstaller folder is already up to date

Starting "rush my-cli"

from my-cli
Try to load  @microsoft/rush-lib from install-run-rush
The Rush engine has not been installed yet. Invoking install-run-rush.js...
Try to load  @microsoft/rush-lib from install-run-rush a second time
```

### Issues not resolved in this PR:

- Remove limitation of `10` in `tryFindRushJsonLocation`
- `RUSH_SDK_DEBUG` blocked by Rush itself... Technically `verboseEnabled` call be easily changed into another variable, so i leave it to you.
- `rush-sdk` logging affacts script logging which may want to logout a JSON string

```
ERROR: The following environment variables were found with the "RUSH_" prefix, but they are not recognized by this version of Rush:
RUSH_SDK_DEBUG
```

**Comment by elliot-nelson** [comment]
For RUSH_SDK_DEBUG, we could add that variable to the list of allowed variables in EnvironmentConfiguration in rush-lib, with a comment that rush-lib doesn't actually use it (rush-sdk does).

This seems more appropriate than using a different non-RUSH prefix, to me.

File: libraries/rush-sdk/src/index.ts
Comment (iclanton) [inline libraries/rush-sdk/src/index.ts:103]:
```suggestion
      terminal.writeVerboseLine(`Trying to load  ${RUSH_LIB_NAME} installed by install-run-rush`);
```

File: libraries/rush-sdk/src/index.ts
Comment (iclanton) [inline libraries/rush-sdk/src/index.ts:126]:
```suggestion
        terminal.writeVerboseLine(`Trying to load  ${RUSH_LIB_NAME} installed by install-run-rush a second time`);
```

File: libraries/rush-sdk/src/index.ts
Comment (iclanton) [inline libraries/rush-sdk/src/index.ts:137]:
```suggestion
      terminal.writeVerboseLine(`Loaded ${RUSH_LIB_NAME} installed by install-run-rush`);
```

**Pull request merged**
